### DISCUSSION TITLE
Distribute flux as a crate?

### DISCUSSION THREAD

**Discussion by Milo123459** [flux-rs/flux#570] [Ideas]
Could flux ever be distributed as a crate? The only thing preventing me from using this is the fact it requires a custom rustc.

**Comment by EDMPhoenix** [comment]
In theory, it's possible. However, there are some practical considerations.

Flux, as it's designed right now, relies on custom Rust compiler modifications to achieve its data-and control-flow analyses, in order to reason about permissions and capabilities, something the standard Rust compiler doesn't offer.

So, as long as flux relies on these special compiler features, it cannot readily be distributed as a standard crate unless the features that flux depends on have a later implementation in stable Rust.

However, as Rust evolves and if these features become part of the stable branch of the language, then it would be possible to distribute it as a crate on https://crates.io/.

**Reply by Milo123459** [reply]
Could the dependence for these special features be changed? But, this makes sense - just slightly annoying haha

**Reply by EDMPhoenix** [reply]
Actually, changing the dependence on these special features may not be easy. Flux is built in a way that heavily relies on these features. It's almost like asking a car to run without an engine - it's very foundational to how Flux operates.

I totally get that it might be annoying. But as Rust continues to develop, there's a chance these unique features that Flux uses might become part of the main Rust language. When that happens, we could then distribute Flux as a regular crate!

**Reply by Milo123459** [reply]
Awesome, thanks for letting me know!

**Reply by EDMPhoenix** [reply]
My pleasure. Please feel free to ask any questions.

**Comment by nilehmann** [comment]
Hi, @Milo123459.  What exactly is the blocker for using Flux? While you need a nightly version of the compiler installed to run Flux, you can still compile and run your code with a (different) stable version. If you use Rustup it should be straightforward to have both versions (granted we don't have good documentation around it). Do you not use Rustup? do you have some other restrictions?

While Flux will probably never be "just a crate" we are interested in improving the way we distribute it. The issue is that Flux has external dependencies that can't be bundled in a single binary and we need to figure out how to distribute those. Unfortunately,  we currently don't have too many cycles to work on it. Help would be appreciated.

**Reply by Milo123459** [reply]
In some environments, rustup won't be available - which is why i was asking
